compiler/rustc_middle/src/mir/syntax.rs

```diff
@@ -180,11 +183,8 @@ pub enum BorrowKind {
     ///     Some(_) => (),
     /// }
     /// ```
-    /// This can't be a shared borrow because mutably borrowing (*x as Some).0
-    /// should not prevent `if let None = x { ... }`, for example, because the
-    /// mutating `(*x as Some).0` can't affect the discriminant of `x`.
     /// We can also report errors with this kind of borrow differently.
-    Fake,
+    Fake(FakeBorrowKind),
```

Comment:
It might make more sense to add a `StatementKind::FakeAssign`, and let `BorrowKind` have variants `Shared`/`ShallowShared`/`Mut`

Reply:
Or even a `remove_after_borrowck: bool` flag on `Statement` tbh

Reply:
Having this as as only a borrow kind is nice for reducing the places that need to handle this special case. I think exactly what kinds of fake borrows we need probably depends on how match checking is handling deref patterns.

Reply:
The issue is that `Fake(Deep)` should behave exactly like `Shared` apart from the fact that it's removed after borrowck. I'm worried future maintainers may not keep that invariant

Reply:
I think the cleanest would be `Rvalue::Ref(.., BorrowKind, BorrowFakeness)`, but that's a big change. I'm only doing that if others think this is a good idea.

Reply:
Re. match checking, the intent is that we'd consider `deref!(true) | deref!(false)` exhaustive since we're only targeting well-behaved std types. So we need fkae borrows to avoid [unsoundness](https://github.com/rust-lang/rust/pull/122598#discussion_r1554347440).

Reply:
OK, I think that this is what we want then